File: vvd-common/src/main/java/mikufan/cx/vvd/common/util/FileNameUtil.java
package mikufan.cx.vvd.common.util;

import mikufan.cx.vvd.common.vocadb.model.SongForApi;

/**
 * @author CX无敌
 * @date 2020-12-19
 */
public interface FileNameUtil {

  static String buildErrorInfoJsonFileName(SongForApi song){
    return buildBasicFileNameForSong(song) + FileNamePostFix.SONG_INFO_ERR + ".json";
  }

  static String buildInfoJsonFileName(SongForApi song){
    return buildBasicFileNameForSong(song) + FileNamePostFix.SONG_INFO + ".json";
  }

  static String buildPvFileName(SongForApi song, String extension){
    return buildBasicFileNameForSong(song) + FileNamePostFix.VIDEO + "." + extension;
  }

  static String buildThumbnailFileName(SongForApi song, String extension){
    return buildBasicFileNameForSong(song) + FileNamePostFix.THUMBNAIL + "." + extension;
  }

  static String buildBasicFileNameForSong(SongForApi song){
    var artists = song.getArtistString().split("feat\\.");
    var vocals = artists[1].trim();
    var producers = artists[0].trim();
    var songName = song.getName();
    return removeIllegalChars(String.format("【%s】%s【%s】", vocals, songName, producers));
  }

  static String removeIllegalChars(String fileName){
    return fileName
        .replace("/", "-")
        .replace("\\", "-")

        .replace("? ", " ")
        .replace("* ", " ")
        .replace(": ", " ")

        .replace("?", " ")
        .replace("*", " ")
        .replace(":", " ");
  }
}

File: vvd-commonkt/src/main/kotlin/mikufan/cx/vvd/commonkt/naming/FileNameUtil.kt
package mikufan.cx.vvd.commonkt.naming

import mikufan.cx.vvd.commonkt.vocadb.api.model.SongForApiContract
import mikufan.cx.vvd.common.naming.FileNamePostFix

/**
 * Contain the basic filename naming for [SongForApiContract]
 *
 * Other naming such as toPvFileName, toLabelFileName, etc. are extension of [SongProperFileName]
 * in order to enforce the common [SongProperFileName] to be reused
 *
 * Such extensions are all in their specific module
 *
 * @author CX无敌
 * @date 2020-12-19
 */

fun SongForApiContract.toErrorFileName(): String {
  return toProperFileName().toString() + FileNamePostFix.ERROR + ".json"
}

/**
 * Take a [SongForApiContract] and generate its proper filename that is confliction-free
 *
 * By default, all filename generation should be drived from this name
 */
fun SongForApiContract.toProperFileName(): SongProperFileName {
  return buildSongProperFileName(artistString, defaultName, id)
}

/**
 * Build a consistent filename structure from common song fields.
 */
private fun buildSongProperFileName(
  artistString: String?,
  defaultName: String?,
  id: Int?
): SongProperFileName {
  val artists = requireNotNull(artistString) { "artist string is null" }.split("feat.")
  require(artists.size >= 2) { "artist string missing 'feat.': $artistString" }
  val vocals = artists[1].trim()
  val producers = artists[0].trim()
  val songName: String = requireNotNull(defaultName) { "song name is null" }
  val vocadbId = requireNotNull(id) { "id is null" }
  return SongProperFileName(removeIllegalChars(String.format("【%s】%s【%s】[%s]", vocals, songName, producers, vocadbId)))
}

fun removeIllegalChars(fileName: String): String {
  return fileName
    .replace("/", "-")
    .replace("\\", "-")
    .replace('"', '\'')
    .replace("? ", " ")
    .replace("* ", " ")
    .replace(": ", " ")
    .replace(Regex("[\\\\/:*?\"<>|]"), " ")
}
